### PR TITLE
allowing to start from selected state

### PR DESCRIPTION
**Pull request by lukeses**
Hi @sasa1977

I've made a change to a function that initializes fsm. What do you think about merging into master this change. My motivation is that I want to persist a state from a database and go directly to the selected state.

### CODE REVIEW

**Comment by OvermindDL1** [comment]
What about the `data` too?

**Comment by sasa1977** [comment]
Thanks for contributing back. I agree with @OvermindDL1 that it would be nice to support the data too. Perhaps something like

```elixir
# initialize with default values
MyFsm.new()

# initialize with a custom state
MyFsm.new(state: foo)

# initialize with a custom data
MyFsm.new(data: bar)

# initialize with a custom state and data
MyFsm.new(state: foo, data: bar)
```

What do you think?

**Comment by OvermindDL1** [comment]
_/me personally just `binary_to_term/2`/`term_to_binary/2` the entire FSM directly, but it would be nice for consistency..._

**Comment by lukeses** [comment]
Thank you for feedback. I've made changes that you propose @sasa1977.

@OvermindDL1 Can you elaborate a little more, when it comes to using `binary_to_term` or `term_to_binary`?

File: lib/fsm.ex [hunk 1]
```diff
@@ -13,10 +13,18 @@ defmodule Fsm do
...
       end
 
+      def new(data: initial_data) do
+        %{%__MODULE__{} | data: initial_data}
+      end
+
+      def new(state: initial_state, data: initial_data) do
+        %{%__MODULE__{} | state: initial_state, data: initial_data}
+      end
+
```
Comment (sasa1977) [inline lib/fsm.ex]: I think you could replace all four definitions of `new` with the single one:

```elixir
def new(params \\ []), do:
  struct!(__MODULE__, params)
```

Could you please give it a try?

**Comment by lukeses** [comment]
@sasa1977 Yes, it works. Thank you for a quick review!

**Comment by sasa1977** [comment]
Thank you for contributing! I'll give it a try myself later today and merge.

**Pull request merged**